packages/tailwindcss/src/variants.ts

```diff
@@ -205,99 +207,138 @@ export function createVariants(theme: Theme): Variants {
   variants.static('force', () => {}, { compounds: false })
   staticVariant('*', [':where(& > *)'], { compounds: false })
 
+  // The `not-*` variant
+  //
+  // This variant only supports inversion of style rules whether nested or not.
+  //
+  // Inversion of at-rules is explicitly not handled because it introduces
+  // signfiicant complexity. Supporting at-rules would be a purely additive
+  // change and as such can be done in the future without breaking changes.
+  //
+  // As a demonstration of the complexity, consider the `hover` variant:
+  // ```css
+  // @media (hover: hover) {
+  //   &:hover {
+  //     /* declarations */
+  //   }
+  // }
+  // ```
+  //
+  // This requires the media query AND the `:hover` selector to match and is
+  // is equvivalent to the following logically:
+  // media hover && selector :hover
+  //
+  // Negating then simplifying by De Morgan's laws results in:
+  // -> not(media hover && selector :hover)
+  // -> not(media hover) || not(selector :hover)
+  // -> media not hover || selector :not(:hover)
+  //
+  // Which means that we now have to act like the following definition:
+  // ```css
+  // @media not (hover: hover) {
+  //   /* declarations */
+  // }
+  //
+  // &:not(:hover) {
+  //   /* declarations */
+  // }
+  // ```
+  //
+  // Further, for a variant using multiple at-rules as an OR we would have to
+  // combine them.
+  //
+  // For example, given the following:
+  //
+  // ```css
+  // @media (hover: hover), (any-hover: fine) {
+  //   &:hover {
+  //     /* declarations */
+  //   }
+  // }
+  // ```
+  //
+  // Which is equivalent to the following logically:
+  // (media hover || media any-hover) && selector :hover
+  //
+  // Negating then simplifying by De Morgan's laws results in:
+  // -> not((media hover || media any-hover) && selector :hover)
+  // -> not(media hover || media any-hover) || not(selector :hover)
+  // -> (media not hover && media not any-hover) || selector :not(:hover)
+  //
+  // Which means that we now have to act like the following definition:
+  // ```css
+  // @media not (hover: hover) {
+  //   @media not (any-hover: fine) {
+  //     /* declarations */
+  //   }
+  // }
+  //
+  // &:not(:hover) {
+  //   /* declarations */
+  // }
+  // ```
+  //
+  // Which as you can see can introduce arbitrary levels of nesting and
+  // complexity. This is why we do not support inverting at-rules.
   variants.compound('not', (ruleNode, variant) => {
```

Comment:
One thing to figure out with this `not` variant is that it _can_ compounds other variants. Right now all `compound()` variants will check if the variant we are nesting have the `compounds: true` property. 

This makes sense for `peer-*`, `group-*`, and `has-*`, since those operate on selectors and most variants with `compound: false` (like breakpoints) don't have selectors (so `group-md` doesn't make sense).

But in this case, for `not`, it does make sense. Because right now you cannot do `not-portrait` or `not-md`.

Reply:
Same goes for custom `@variant` definitions in your CSS. They are marked as `compounds: false`.